### PR TITLE
LNX-188 Add Drop Action

### PR DESCRIPTION
**Pull request by EasyJB**

### CODE REVIEW

**Review by Krystian030 — APPROVED** [review]

File: lynx/common/actions/common_requirements.py [hunk 1]
```diff
@@ -70,3 +70,6 @@ def any_object_on_square_has_all_given_tags(scene: Scene, position: Vector, give
                 return True
         return False
 
+    def has_something_in_inventory(scene: 'Scene', object_id: int) -> bool:
+        object: Object = scene.get_object_by_id(object_id)
+        return(bool(object.inventory))
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]: `return bool(object.inventory)`

File: lynx/common/actions/common_requirements.py [hunk 2]
```diff
@@ -70,3 +70,6 @@ def any_object_on_square_has_all_given_tags(scene: Scene, position: Vector, give
                 return True
         return False
 
+    def has_something_in_inventory(scene: 'Scene', object_id: int) -> bool:
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py:73]: `has_object_with_given_id_in_inventory`

File: lynx/common/actions/drop.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+from lynx.common.actions.create_object import CreateObject
+from lynx.common.actions.update_resources import UpdateResources
+from lynx.common.object import Object
+from lynx.common.vector import Vector
+
+
+@dataclass
+class Drop(Action):
+    """
+    Action used to empty inventory of agent.
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py]: `agent` → `Agent`

File: lynx/common/actions/drop.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from dataclasses import dataclass
+from collections import defaultdict
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py]: I don't see this used anywhere

File: lynx/common/actions/drop.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+from lynx.common.object import Object
+from lynx.common.vector import Vector
+
+
+@dataclass
+class Drop(Action):
+    """
+    Action used to empty inventory of agent.
+    If target_position is equal to drop_area_position, we should increase global points and resources.
+    If target_position is different than drop_area_position then we should create each object from inventory on choosen
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py]: `choosen` → `chosen`

File: lynx/common/actions/drop.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    square (target_position)
+    """
+    object_id: int = -1
+    target_position: Vector = Vector(0, 1)
+
+    def apply(self, scene: 'Scene') -> None:
+        agent: Object = scene.get_object_by_id(self.object_id)
+        player_name = agent.owner
+        if self.target_position == scene.get_drop_area_of_a_player(player_name):
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py:41]: Definitely create helper function for these 2 cases e.g. `drop_in_drop_area`, `drop_in_overworld`

File: lynx/common/actions/drop.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            scene.update_resources_of_player(player_name, agent.inventory)
+            updated_points_action = UpdateResources(player_name, agent.inventory)
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py]: Does the `Action` update points or resources? There's a conflict here.

File: lynx/common/actions/update_resources.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+from typing import Dict
+from dataclasses import field
+
+from lynx.common.actions.action import Action
+
+
+@dataclass
+class UpdateResources(Action):
+    """
+    Simple action for updating points on fronted.
```
Comment (filip-szweda) [inline lynx/common/actions/update_resources.py]: Again, points or resources? Also, `frontend`

File: lynx/common/actions/update_resources.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+from dataclasses import dataclass
+from typing import Dict
+from dataclasses import field
+
+from lynx.common.actions.action import Action
+
+
+@dataclass
+class UpdateResources(Action):
```
Comment (filip-szweda) [inline lynx/common/actions/update_resources.py:9]: If some `Player` joins mid-game, they won't have any resources/points to put in their scoreboard until someone `Drops` resources in `DropArea`.

IMHO `UpdateResources` should be added to pending `Actions` upon new player joining. An alternative is adding resources/points to standard `GET` request answer, but that would be too heavy on the server.

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: lynx/common/actions/common_requirements.py (hunk 2 above)
Comment (EasyJB) [reply lynx/common/actions/common_requirements.py:73]: It implicates that we check if given object is in inventory.

File: lynx/common/actions/update_resources.py (hunk 9 above)
Comment (EasyJB) [reply lynx/common/actions/update_resources.py:9]: In general, in games, you should see what resources you can have (even if you don't have any)

File: lynx/common/actions/drop.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    Action used to empty inventory of Agent.
+    If target_position is equal to drop_area_position, we should increase global points and resources.
+    If target_position is different than drop_area_position then we should create each object from inventory on chosen
+    square (target_position)
+    """
+    object_id: int = -1
+    target_position: Vector = Vector(0, 1)
+
+    def drop_in_drop_area(self, scene: 'Scene', player_name: str, inventory: dict):
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py:22]: `-> None`

File: lynx/common/actions/drop.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        scene.update_resources_of_player(player_name, inventory)
+        update_resources_action = UpdateResources(player_name, inventory)
+        scene.add_to_pending_actions(update_resources_action.serialize())
+
+    def drop_in_overworld(self, scene: 'Scene', inventory: dict):
```
Comment (filip-szweda) [inline lynx/common/actions/drop.py:27]: `-> None`

File: lynx/common/scene.py [hunk 12]
```diff
@@ -88,3 +90,22 @@ def generate_drop_area(self, player: str) -> None:
...
+                return player
+        return None
+
+    def get_drop_area_of_a_player(self, player_id: str) -> Optional[Vector]:
+        for player in self.players:
+            if player.player_id == player_id:
+                return player.drop_area
+        return None
+
+    def update_resources_of_player(self, player_id: str, inventory: Dict):
```
Comment (filip-szweda) [inline lynx/common/scene.py:108]: `-> None`

File: tests/drop_test.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        expected_scene = Scene(players=[Player(player_id="dummy", player_resources={"Wood": 0, "Stone": 0}, drop_area=Vector(5, 5))])
+        expected_dummy_object = Object(id=1, name="dummy", owner="dummy", position=Vector(5, 5))
+        expected_dummy_drop = Drop(target_position=Vector(5, 6), object_id=1)
+        expected_dummy_object2 = Object(id=expected_scene.generate_id(), name="Wood", position=Vector(5, 6),
+                                        tags=['pushable', 'pickable'])
+        expected_scene.add_entity(expected_dummy_object)
+        expected_scene.add_entity(expected_dummy_drop)
+        expected_scene.add_entity(expected_dummy_object2)
+
+        random.seed(1222)
```
Comment (filip-szweda) [inline tests/drop_test.py:39]: Do we really need to call it twice in one function? Call it in-class e.g.
```
class TestDropApply:
   random.seed(1222)
```

File: tests/drop_test.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+        scene = Scene(players=[Player(player_id="test", player_resources={"Wood": 0, "Stone": 0}, drop_area=Vector(5, 6))])
+        dummy_object = Object(id=1, name="dummy", owner="test", position=Vector(5, 5), inventory={"Wood": 2, "Stone": 1})
+        dummy_drop = Drop(target_position=Vector(5, 6), object_id=1)
+        scene.add_entity(dummy_object)
+        scene.add_entity(dummy_drop)
+
+        dummy_drop.apply(scene)
+
+        assert scene == expected_scene
```
Comment (filip-szweda) [inline tests/drop_test.py:101]:
```
assert scene == expected_scene

class TestDropRequirements:
   def test_success_requirements(self):
```

File: tests/drop_test.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+    def test_success_requirements(self):
+        scene = Scene(players=[Player(player_id="dummy", player_resources={"Wood": 0, "Stone": 0}, drop_area=Vector(5, 5))])
+        dummy_object = Object(id=1, name="dummy", owner="dummy", position=Vector(5, 5), inventory={"Wood": 1})
+        dummy_drop = Drop(target_position=Vector(5, 6), object_id=1)
+        scene.add_entity(Object(id=3, name="Grass", position=Vector(5, 6), tags=['walkable']))
+        scene.add_entity(dummy_object)
+        assert dummy_drop.satisfies_requirements(scene) is True
+
+    def test_fail_requirements_distance(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: This test actually checks what happens when there's no `Tile` to `Drop` on
`test_no_tile_to_drop_on_negative/fails`

File: tests/drop_test.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+from lynx.common.object import Object
+from lynx.common.player import Player
+from lynx.common.scene import Scene
+from lynx.common.vector import Vector
+from lynx.common.actions.update_resources import UpdateResources
+
+class TestDropSerialization:
+    expected_serialization_drop = '{"type": "Drop", "attributes": {"object_id": 1, "target_position": {"x": 1, "y": 0}}}'
+
+    def test_success_serialization(self) -> NoReturn:
```
Comment (filip-szweda) [inline tests/drop_test.py]: `-> None` everywhere in this file

File: tests/drop_test.py [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+        dummy_drop = Drop(target_position=Vector(5, 6), object_id=1)
+        scene.add_entity(dummy_object)
+        scene.add_entity(dummy_drop)
+
+        dummy_drop.apply(scene)
+
+        assert scene == expected_scene
+class TestDropRequirements:
+
+    def test_success_requirements(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: `test_all_satisified_positive/success`

File: tests/drop_test.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+    def test_fail_requirements_distance(self):
+        scene = Scene(players=[Player(player_id="dummy", player_resources={"Wood": 0, "Stone": 0}, drop_area=Vector(5, 5))])
+        dummy_object = Object(id=1, name="dummy", owner="dummy", position=Vector(5, 5), inventory={"Wood": 1})
+        dummy_drop = Drop(target_position=Vector(6, 6), object_id=1)
+        scene.add_entity(Object(id=3, name="Grass", position=Vector(5, 6), tags=['walkable']))
+        scene.add_entity(dummy_object)
+        assert dummy_drop.satisfies_requirements(scene) is not True
+
+    def test_fail_requirements_inventory(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: `test_object_with_no_inventory_negative/fails`

File: tests/drop_test.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+        scene = Scene(players=[Player(player_id="dummy", player_resources={"Wood": 0, "Stone": 0}, drop_area=Vector(5, 5))])
+        dummy_object = Object(id=1, name="dummy", owner="dummy", position=Vector(5, 5))
+        dummy_drop = Drop(target_position=Vector(5, 6), object_id=1)
+        scene.add_entity(Object(id=3, name="Grass", position=Vector(5, 6), tags=['walkable']))
+        scene.add_entity(dummy_object)
+        assert dummy_drop.satisfies_requirements(scene) is not True
+
+    def test_fail_requirements_walkable(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: `test_no_walkable_tile_to_drop_on_negative/fails`

File: tests/user_helper_functions_test.py [hunk 20]
```diff
@@ -50,6 +50,7 @@ def test_filter_objects_failure(self) -> NoReturn:
         assert filter_objects(self.scene, [1, 2, 3, 4], "empty") != [1]
 
     def test_random_direction_success(self) -> NoReturn:
+        random.seed(12)
```
Comment (filip-szweda) [inline tests/user_helper_functions_test.py:52]: Already done in line 20

File: tests/drop_test.py [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+        assert self.expected_serialization_drop == serialized_drop
+
+    def test_success_deserialization(self):
+        expected_drop = Drop(object_id=1, target_position=Vector(1, 0))
+        dummy_drop = Drop.deserialize(self.expected_serialization_drop)
+
+        assert dummy_drop == expected_drop
+
+class TestDropApply:
+    def test_success_apply(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: Please explain test scenarios in names more throughly e.g.
`test_drop_single_resource_in_overworld_creates_single_object`

File: tests/drop_test.py [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+        scene.add_entity(dummy_drop)
+
+        dummy_drop.apply(scene)
+        for action in scene.pending_actions:
+            Action.deserialize(action).apply(scene)
+        scene.pending_actions.clear()
+
+        assert scene == expected_scene
+
+    def test_success_multiple_objects_apply(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: `test_drop_multiple_resources_in_overworld_creates_multiple_objects`

File: tests/drop_test.py [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+        scene.add_entity(dummy_drop)
+
+        dummy_drop.apply(scene)
+        for action in scene.pending_actions:
+            Action.deserialize(action).apply(scene)
+        scene.pending_actions.clear()
+
+        assert scene == expected_scene
+
+    def test_success_drop_to_drop_area_apply(self):
```
Comment (filip-szweda) [inline tests/drop_test.py]: `test_drop_multiple_resources_in_drop_area_adds_multiple_resources_to_player`

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: tests/drop_test.py (hunk 13 above)
Comment (EasyJB) [reply tests/drop_test.py:39]: Yes, because we ned to reset seed. Without this test will fail.

**Review by filip-szweda — APPROVED** [review]

File: tests/drop_test.py [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+        scene.add_entity(dummy_object)
+        scene.add_entity(dummy_drop)
+
+        dummy_drop.apply(scene)
+
+        assert scene == expected_scene
+
+
+class TestDropRequirements:
+
```
Comment (filip-szweda) [inline tests/drop_test.py:105]: No need for this empty line

File: tests/drop_test.py [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    def test_all_requirements_satisified_positive(self) -> None:
```
Comment (filip-szweda) [inline tests/drop_test.py:106]: You use `fail` in this class, so maybe `success`?

**Pull request merged**